echopype/convert/parse_base.py

```diff
@@ -69,54 +76,274 @@ def _print_status(self):
+            ping_data_dict = (
+                self.ping_data_dict_tx if raw_type == "transmit" else self.ping_data_dict
+            )
+            for data_type, shape in expanded_shapes.items():
+                if shape:
+                    # Get current data size
+                    size = sum([sys.getsizeof(val) for val in ping_data_dict[data_type].values()])
+                    current_data_size += size
+
+                    # Estimate expansion sizes
+                    itemsize = np.dtype("float64").itemsize
+                    req_mem = np.prod(shape) * itemsize
+                    total_req_mem += req_mem
+
+        # get statistics about system memory usage
+        mem = psutil.virtual_memory()
+        # approx. the amount of memory that will be used after expansion
+        req_mem = mem.used - current_data_size + total_req_mem
+
+        return mem.total * mem_mult < req_mem
+
+    def rectangularize_data(
+        self,
+        use_swap: "bool | Literal['auto']" = "auto",
+        max_chunk_size: str = "100MB",
+    ) -> None:
         """
         Rectangularize the power, angle, and complex data.
         Additionally, convert the data to a numpy array
         indexed by channel.
         """
+        # Determine use_swap
+        if use_swap == "auto":
+            use_swap = self.__should_use_swap()
+
+        # Perform rectangularization
+        zarr_root = None
+        if use_swap:
+            # Setup temp store
+            zarr_store = create_temp_zarr_store()
+            # Setup zarr store
+            zarr_root = zarr.group(
+                store=zarr_store, overwrite=True, synchronizer=zarr.ThreadSynchronizer()
+            )
 
-        # append zarr datagrams to channel ping data
-        for dgram in self.zarr_datagrams:
-            self._append_channel_ping_data(dgram, zarr_vars=False)
-
-        # append zarr transmit datagrams to channel ping data
-        for dgram in self.zarr_tx_datagrams:
-            self._append_channel_ping_data(dgram, rx=False, zarr_vars=False)
-
-        # Rectangularize all data and convert to numpy array indexed by channel
-        for data_type in ["power", "angle", "complex"]:
-            # Receive data
-            for k, v in self.ping_data_dict[data_type].items():
-                if all(
-                    (x is None) or (x.size == 0) for x in v
-                ):  # if no data in a particular channel
-                    self.ping_data_dict[data_type][k] = None
-                else:
-                    # Sort complex and power/angle channels and pad NaN
-                    self.ch_ids[data_type].append(k)
-                    self.ping_data_dict[data_type][k] = self.pad_shorter_ping(v)
-            # Transmit data
-            self.rectangularize_transmit_ping_data(data_type)
+        # Compute the final expansion shapes for each data type
+        expanded_data_shapes = self._get_data_shapes()
```

Comment:
`self._get_data_shapes()` is also used in `self.__should_use_swap()`, so `expanded_data_shapes` could be calculated first and passed in to `__should_use_swap` as an argument.